### DISCUSSION TITLE
how to generate dsk/bin files?

### DISCUSSION THREAD

**Discussion by therealansh** [vinc/moros#487] [General]
I am trying to build a toy OS using rust and wanted to know how did you generate custom reboot/halt binaries? Did you build it from source cause i can see you adding something like `MOROS has come to its fate...`, How were you able to achieve this?

**Comment by therealansh** [comment]
Also adding to this discussion, how to add additional syscalls? For example, lets say i have another CLI made in Rust, how to use it in MOROS? I assume it can be done by putting it in the `dsk/bin`?

**Reply by vinc** [reply]
Hi, I moved the issue into a discussion!

So first of all the userspace is a work in progress, currently processes don't have a dedicated page table and can't allocate memory from the heap space without crashing, but I implemented the first syscalls a while ago and it's already possible to do interesting things with them.

MOROS can load ELF binaries written in Rust (in `src/bin`) or flat binaries written in assembler (in `dsk/src/bin`). You can compile them with `make user-rust` or `make user-nasm` and the binaries will go to `dsk/bin`. You'll need to strip the rust binaries to make them smaller. And if you want to experiment with new binaries you can update `src/usr/install.rs` to install them in the system.

I would not recommend investing too much time into writing a big program yet because the userspace will still evolve. But its simplicity make it interesting to do little experiments.

For the `halt` command for example I have a `STOP` syscall that takes an argument to either shutdown (`0xdead`) or reboot (`0xcafe`) the system. The syscall dispatcher will call the `stop` service with the argument, and in case of shutdown the service will call a function in the ACPI module.
